### PR TITLE
Return (response, error)

### PR DESCRIPTION
**Pull request by tamird**
Some work toward #1400.

### CODE REVIEW

File: storage/range.go [hunk 1]
```diff
@@ -95,7 +95,7 @@ var (
 // consistent results each time. Should only be used in tests in the
 // storage package but needs to be exported due to circular import
 // issues.
```
Comment (bdarnell) [inline storage/range.go]: This comment needs to be updated.

File: storage/range.go [hunk 2]
```diff
@@ -614,7 +614,10 @@ func (r *Range) addReadOnlyCmd(ctx context.Context, args proto.Request, reply pr
 		if header.Timestamp.Equal(proto.ZeroTimestamp) {
 			header.Timestamp = r.rm.Clock().Now()
 		}
-		intents, err := r.executeCmd(r.rm.Engine(), nil, args, reply)
+		executedReply, intents, err := r.executeCmd(r.rm.Engine(), nil, args)
+		executedReply.Header().SetGoError(err)
```
Comment (bdarnell) [inline storage/range.go]: Before calling `SetGoError` we should assert that `executedReply.Header().GoError()` is nil.

**Comment by mrtracy** [comment]
LGTM, although you might want Ben or Tobias to take a look. From my humble perspective, this does seem much better than using `SetGoError()` directly wherever errors occur.

File: storage/range_command.go [hunk 3]
```diff
@@ -39,59 +39,100 @@ import (
 // executeCmd switches over the method and multiplexes to execute the
 // appropriate storage API command. It returns an error and, for some calls
 // such as inconsistent reads, the intents they skipped.
-func (r *Range) executeCmd(batch engine.Engine, ms *engine.MVCCStats, args proto.Request, reply proto.Response) ([]proto.Intent, error) {
+func (r *Range) executeCmd(batch engine.Engine, ms *engine.MVCCStats, args proto.Request) (reply proto.Response, intents []proto.Intent, err error) {
+	// Ugh. Our callers expect non-nil replies even in the error case.
```
Comment (bdarnell) [inline storage/range_command.go]: Make this a TODO? Which callers require this?

File: storage/range_command.go [hunk 4]
```diff
@@ -39,59 +39,100 @@ import (
...
+
 	// Verify key is contained within range here to catch any range split
 	// or merge activity.
 	header := args.Header()
 
-	if err := r.checkCmdHeader(header); err != nil {
-		reply.Header().SetGoError(err)
-		return nil, err
+	if err = r.checkCmdHeader(header); err != nil {
+		return
```
Comment (bdarnell) [inline storage/range_command.go]: I think we prefer `return nil, nil, err` instead of relying on named return values.

File: storage/range_command.go [hunk 5]
```diff
@@ -118,77 +157,70 @@ func (r *Range) executeCmd(batch engine.Engine, ms *engine.MVCCStats, args proto
...
 
 // Get returns the value for a specified key.
-func (r *Range) Get(batch engine.Engine, args *proto.GetRequest, reply *proto.GetResponse) []proto.Intent {
-	val, intents, err := engine.MVCCGet(batch, args.Key, args.Timestamp, args.ReadConsistency == proto.CONSISTENT, args.Txn)
-	reply.Value = val
-	reply.SetGoError(err)
-	return intents
+func (r *Range) Get(batch engine.Engine, args *proto.GetRequest) (reply proto.GetResponse, intents []proto.Intent, err error) {
+	reply.Value, intents, err = engine.MVCCGet(batch, args.Key, args.Timestamp, args.ReadConsistency == proto.CONSISTENT, args.Txn)
+	return
```
Comment (bdarnell) [inline storage/range_command.go]: I'd rather not use named return values in any of these methods.

File: storage/range_command.go [hunk 6]
```diff
@@ -234,7 +263,7 @@ func (r *Range) EndTransaction(batch engine.Engine, ms *engine.MVCCStats, args *
 		}
 	} else {
 		// The transaction doesn't exist yet on disk; use the supplied version.
-		reply.Txn = gogoproto.Clone(args.Txn).(*proto.Transaction)
+		reply.Txn = args.Txn
```
Comment (bdarnell) [inline storage/range_command.go]: Why did you remove the clone? It looks deliberate (although if it was important I'd hope we have a test for it).

File: storage/range_command.go [hunk 7]
```diff
@@ -412,33 +438,29 @@ func (r *Range) InternalRangeLookup(batch engine.Engine, args *proto.InternalRan
...
-	// If no existing transaction record was found, initialize
-	// to the transaction in the request header.
-	if !ok {
-		gogoproto.Merge(&txn, args.Txn)
+	if ok, err := engine.MVCCGetProto(batch, key, proto.ZeroTimestamp, true, nil, &txn); err != nil {
+		return reply, err
+	} else if !ok {
+		// If no existing transaction record was found, initialize
+		// to the transaction in the request header.
+		txn = *args.Txn
```
Comment (bdarnell) [inline storage/range_command.go]: This is a shallow copy; the old version was a deep copy. I'd rather keep the `Merge` just to be safe.

**Comment by bdarnell** [comment]
LGTM

File: storage/range_command.go (hunk 6 above)
Comment (tamird) [inline storage/range_command.go]: Didn't look like it mattered, and none of the tests complained.

File: storage/range_command.go (hunk 7 above)
Comment (tamird) [inline storage/range_command.go]: I don't see a reason to copy here at all, so I removed the copy altogether.

File: storage/range_command.go (hunk 3 above)
Comment (tamird) [inline storage/range_command.go]: done

File: storage/range.go (hunk 1 above)
Comment (tamird) [inline storage/range.go]: done

File: storage/range.go (hunk 2 above)
Comment (tamird) [inline storage/range.go]: I don't think that's necessary since this code is in transition. I've added the assertion.

File: storage/range_command.go (hunk 5 above)
Comment (tamird) [inline storage/range_command.go]: +1, and I wish we had a linter for this, but it's such an incredible pain to change that until such a linter arrives I'd rather leave this as-is.

Comment (bdarnell) [inline storage/range_command.go]: I don't understand. "As-is" was not using named return values; you just introduced them in this change.

Comment (tamird) [inline storage/range_command.go]: Heh, "As-is" was actually not using any returns at all. I went with named returns because the alternative was adding a `reply := proto.FooResponse{}` at the beginning of each method, and the `return` style was a smaller change to the existing `SetGoError()` style.

Comment (bdarnell) [inline storage/range_command.go]: Yes, I understand it introduces an extra line or two to declare the reply object. I think that's worth doing and I don't think the existence or non-existence of a linter is relevant. Even if there were a linter that could report on named return values I wouldn't want to use it in `make check` because I do think they're appropriate in some cases.

Comment (tamird) [inline storage/range_command.go]: I yield. PTAL.

File: storage/range_command.go [hunk 8]
```diff
@@ -39,59 +39,103 @@ import (
 // executeCmd switches over the method and multiplexes to execute the
 // appropriate storage API command. It returns an error and, for some calls
 // such as inconsistent reads, the intents they skipped.
-func (r *Range) executeCmd(batch engine.Engine, ms *engine.MVCCStats, args proto.Request, reply proto.Response) ([]proto.Intent, error) {
+func (r *Range) executeCmd(batch engine.Engine, ms *engine.MVCCStats, args proto.Request) (proto.Response, []proto.Intent, error) {
+	// TODO(tamird): Remove this.
+	var reply proto.Response
```
Comment (bdarnell) [inline storage/range_command.go]: Does this work? Modifying a return value from a defer is one of those cases where named return values are necessary. I would keep the names on the return values but use `return nil, nil, err` instead of bare `return` (because bare `return` can have surprising results if `err` is shadowed)

Comment (tamird) [inline storage/range_command.go]: derp, i fixed this locally and didn't push. PTAL

**Comment by bdarnell** [comment]
LGTM

File: storage/range.go [hunk 9]
```diff
@@ -614,7 +613,13 @@ func (r *Range) addReadOnlyCmd(ctx context.Context, args proto.Request, reply pr
 		if header.Timestamp.Equal(proto.ZeroTimestamp) {
 			header.Timestamp = r.rm.Clock().Now()
 		}
-		intents, err := r.executeCmd(r.rm.Engine(), nil, args, reply)
+		executedReply, intents, err := r.executeCmd(r.rm.Engine(), nil, args)
+		if executedReply.Header().Error != nil {
+			panic("the world is on fire")
```
Comment (tschottdorf) [inline storage/range.go]: nice message. Though not helpful to outsiders.

File: storage/range.go [hunk 10]
```diff
@@ -614,7 +613,13 @@ func (r *Range) addReadOnlyCmd(ctx context.Context, args proto.Request, reply pr
...
+		}
+		executedReply.Header().SetGoError(err)
+		reply.Reset()
+		gogoproto.Merge(reply, executedReply)
```
Comment (tschottdorf) [inline storage/range.go]: some overhead here and the escape analysis for `executedReply` likely won't look good. What do you win by doing this?

File: storage/range.go [hunk 11]
```diff
@@ -910,21 +919,26 @@ func (r *Range) applyRaftCommandInBatch(ctx context.Context, index uint64, origi
...
 			// We successfully read from the response cache, so return whatever error
 			// was present in the cached entry (if any).
-			return batch, reply.Header().GoError()
+			return batch, cachedReply.Header().GoError()
 		}
 	}
 
 	// Execute the command.
-	intents, rErr := r.executeCmd(batch, ms, args, reply)
+	executedReply, intents, rErr := r.executeCmd(batch, ms, args)
```
Comment (tschottdorf) [inline storage/range.go]: no "world on fire" assertion?

File: storage/range_command.go [hunk 12]
```diff
@@ -39,59 +39,98 @@ import (
 // executeCmd switches over the method and multiplexes to execute the
 // appropriate storage API command. It returns an error and, for some calls
 // such as inconsistent reads, the intents they skipped.
```
Comment (tschottdorf) [inline storage/range_command.go]: comment out of date.

File: storage/range_command.go (hunk 6 above)
Comment (tschottdorf) [inline storage/range_command.go]: It's likely that this doesn't matter but it's implicit that methods shouldn't change the arguments unless they consciously chose to do that. Here we're setting two pointers equal and then changing away. If there's ever a bug here, it will be painful to find. Let's just keep the copy.

Comment (tamird) [inline storage/range_command.go]: How would you feel about this method taking its argument by value?

File: storage/range.go (hunk 9 above)
Comment (tamird) [inline storage/range.go]: Yeah, this message sucks, but it's pretty transient as I move this code up the stack.

File: storage/range.go (hunk 10 above)
Comment (tamird) [inline storage/range.go]: It's totally required since `reply` is passed in and mutated.

File: storage/range.go (hunk 11 above)
Comment (tamird) [inline storage/range.go]: probably just missed it. this is already gone in the next PR (which is merged)

File: storage/range_command.go (hunk 6 above)
Comment (tamird) [inline storage/range_command.go]: bah, that still gives you a shallow copy.

Comment (tschottdorf) [inline storage/range_command.go]: yes. Can you put the merge back in with a comment to reevaluate whether we need it? Just so that I can sleep at night.
